### PR TITLE
source/extensions/filters/http/response_map: support disabling using per route config

### PR DESCRIPTION
**Pull request by esmet**
Copies the pattern in source/extensions/filters/http/grpc_http1_reverse_bridge/*

### CODE REVIEW

**Comment by LukeShu** [comment]
> Copies the pattern in source/extensions/filters/http/grpc_http1_reverse_bridge/*

That would be a good thing to put in the commit message, FWIW.

File: api/envoy/extensions/filters/http/response_map/v3/response_map.proto [hunk 1]
```diff
@@ -81,3 +81,14 @@ message ResponseMap {
...
 }
+
+// Extra settings on a per virtualhost/route/weighted-cluster level.
+message ResponseMapPerRoute {
+  oneof override {
+    option (validate.required) = true;
+
+    // Disable the response map filter for this particular vhost or route.
+    // If disabled is specified in multiple per-filter-configs, the most specific one will be used.
+    bool disabled = 1 [(validate.rules).bool = {const: true}];
```
Comment (LukeShu) [inline api/envoy/extensions/filters/http/response_map/v3/response_map.proto:92]: I see those `validate.` options on a few other filters (but not the `grpc_http1_reverse_bridge` filter that you mentioned cribbing the approach from), so I guess I trust that they're correct, but I'm flummoxed as to interpreting what they actually mean.

File: source/extensions/filters/http/response_map/config.cc [hunk 2]
```diff
@@ -19,6 +19,14 @@ Http::FilterFactoryCb ResponseMapFilterFactory::createFilterFactoryFromProtoType
...
 }
 
+Router::RouteSpecificFilterConfigConstSharedPtr
+ResponseMapFilterFactory::createRouteSpecificFilterConfigTyped(
+    const envoy::extensions::filters::http::response_map::v3::ResponseMapPerRoute&
+        proto_config,
+    Server::Configuration::ServerFactoryContext&, ProtobufMessage::ValidationVisitor&) {
+  return std::make_shared<FilterConfigPerRoute>(proto_config);
+}
+
```
Comment (LukeShu) [inline source/extensions/filters/http/response_map/config.cc]: As simple as this block is, I found it hard to read.  I guess that's just to be expected when you have type names that are 71 characters log.  And I guess that's just to be expected with C++/Envoy :/ 

Does 1 arg per line make it better?

```C++
Router::RouteSpecificFilterConfigConstSharedPtr
ResponseMapFilterFactory::createRouteSpecificFilterConfigTyped(
    const envoy::extensions::filters::http::response_map::v3::ResponseMapPerRoute& proto_config,
    Server::Configuration::ServerFactoryContext&,
    ProtobufMessage::ValidationVisitor&) {
  return std::make_shared<FilterConfigPerRoute>(proto_config);
}
```

Maybe? But does the long `proto_config` line run afoul of some automated style checker?

File: source/extensions/filters/http/response_map/response_map_filter.cc [hunk 3]
```diff
@@ -26,14 +29,35 @@ ResponseMapFilter::ResponseMapFilter(ResponseMapFilterConfigSharedPtr config)
...
+
   request_headers_ = &request_headers;
   return Http::FilterHeadersStatus::Continue;
 }
 
 Http::FilterHeadersStatus ResponseMapFilter::encodeHeaders(Http::ResponseHeaderMap& headers,
                                                            bool end_stream) {
-  ENVOY_LOG(trace, "response map filter: encodeHeaders with http status = {} and end_stream = {}",
-      headers.getStatusValue(), end_stream);
+  ENVOY_LOG(trace, "response map filter: encodeHeaders with http status = {}, end_stream = {}, disabled = {}",
```
Comment (LukeShu) [inline source/extensions/filters/http/response_map/response_map_filter.cc]: Should that be `, and disabled =`? IDK what style is, I'm just noting that there was an `and` there before.

**Review by LukeShu — APPROVED** [review]
lgtm, but I do want to understand the protobuf options.

File: source/extensions/filters/http/response_map/response_map_filter.cc (hunk 3 above)
Comment (esmet) [reply source/extensions/filters/http/response_map/response_map_filter.cc]: i removed the `and` in favor of comma separating the 3 values

File: api/envoy/extensions/filters/http/response_map/v3/response_map.proto (hunk 1 above)
Comment (esmet) [reply api/envoy/extensions/filters/http/response_map/v3/response_map.proto:92]: i tested this and it means disabled must be a bool equal to true. false is not valid.

Comment (LukeShu) [reply api/envoy/extensions/filters/http/response_map/v3/response_map.proto:92]: Why shouldn't false be valid? And if it's `validate.required`, then how do you omit it (I guess you just leave out the entire `ResponseMapPerRoute`?)?

Comment (esmet) [reply api/envoy/extensions/filters/http/response_map/v3/response_map.proto:92]: i think the idea is to omit if you want it to be not disabled. it's opinionated for sure.

**Review by kflynn — APPROVED** [review]
Looks OK, I think.

**Pull request merged**